### PR TITLE
docs(assistant): add Prime Intellect model flight log

### PR DESCRIPTION
**Pull request by kevinmichaelchen**
## Summary

- add a durable Beanline Prime Intellect flight log for model-training decisions and eval results
- document the Qwen 0.8B training arc: vanilla baseline, Iteration 1, Iteration 2 champion, price-format hardening, receipt-style follow-up experiments, and supervised receipt artifact prep
- capture the value story: the champion model preserves operational order correctness while reducing repeated tool use and turn drift
- record negative results too, including 0.8B receipt hardening, raw 2B/4B probes, and a stopped 2B training probe

## Key results recorded

Initial vanilla -> Iteration 1:
- reward avg: 0.417 -> 0.775
- tool correctness: 0.484 -> 0.906
- pass@1: 0.562 -> 0.812

Iteration 1 strict baseline -> Iteration 2 champion:
- reward avg: 0.703 -> 0.907
- tool correctness: 0.953 -> 1.000
- final response quality: 0.286 -> 0.729
- product efficiency: 0.536 -> 0.875
- average tool calls: 3.000 -> 1.812
- max-turn rate: 0.562 -> 0.000

Current champion on explicit hard holdout (`0.1.5`, `split=hard_eval`):
- reward avg: 0.830
- pass@1/pass@2: 1.000 / 1.000
- tool correctness: 0.984
- price format: 0.625
- receipt style: 0.866
- average tool calls: 1.750

Rejected follow-up experiments:
- 0.8B receipt-style warm start final trainer hard eval: 0.743, below champion 0.830
- raw 2B hard eval: 0.610, better price prior but weak tool correctness
- raw 4B hard eval: 0.674, verbose/truncated
- 2B training probe: 0.668 at first trained hard eval, stopped to protect budget

Supervised receipt artifact:
- generated `.context/prime-intellect/lab/data/effect_coffee_sft/receipt_final_response_sft.jsonl`
- generated `.context/prime-intellect/lab/data/effect_coffee_sft/receipt_tool_trajectory_sft.jsonl`
- 22 rows each, validated for exact `$x.xx` receipt totals and no cents/wrong-currency wording
- no further RL spend started; current lab lacks the `uv run sft` entrypoint

Budget:
- Current wallet balance: $43.27 from the initial $50
- Approximate total spend: $6.73

## Why this matters

The champion adapter is better at obeying the coffee shop's operational rules: it checks the menu, uses canonical tool arguments, avoids bad order submissions, and stops instead of looping through repeated tools. The hard holdout now gives us a clearer product gate: a model must keep the Iteration 2 correctness and efficiency gains while improving exact receipt totals in dollars.

The latest work says bigger RL is not the next best spend. Bigger raw Qwen models have somewhat better receipt priors, but they do not naturally preserve our required tool discipline; the 2B training probe did not close the gap cheaply. The next step should be running the deterministic receipt corpus through an SFT-capable runtime, or converting it into a receipt-drill warmup only if SFT remains unavailable.

## Tests

- commit hooks passed: oxfmt check, oxlint, lintcn, typecheck

### CODE REVIEW

**Comment by kevinmichaelchen** [comment]
This change is part of the following stack:

- #45 ◀

<sub>Change managed by [git-spice](https://abhinav.github.io/git-spice/).</sub>

**Comment by kevinmichaelchen** [comment]
**ELI5 Version**

We taught a small Qwen model to be a better coffee-shop ordering assistant.

Think of it like training a cashier:
- The raw model is a new cashier who kind of guesses.
- We gave it a menu, order tools, and examples.
- We graded whether it used the right tool, placed valid orders, refused impossible drinks, and gave a useful final receipt.
- We trained it with Prime Intellect until it became better than the raw model.

<details>
<summary>👉 More info</summary>

Current champion:
`Qwen/Qwen3.5-0.8B:rqvfrcdy4xt95oka37b0xjyu`

It is not currently deployed, but we can redeploy it.

</details>

**What We Did**

<details>
<summary>👉 More info</summary>

1. Built a custom Prime Intellect environment for Beanline coffee ordering.
2. Trained a small `Qwen/Qwen3.5-0.8B` adapter.
3. Improved it across two successful iterations.
4. Added a harder holdout test for receipt/price formatting.
5. Tried two more price-format hardening runs.
6. Rejected those later runs because they were worse than the current champion.
7. Kept cost low: wallet is now `$45.94` from the original `$50`.

</details>

**How We Grade It**

<details>
<summary>👉 More info</summary>

We grade it with eval tasks, like a report card.

Main grades:
- `tool_correctness`: Did it call the right tool with the right values?
- `final_response_quality`: Did it answer clearly and mention the right things?
- `product_efficiency`: Did it avoid wasting tool calls or looping?
- `price_format`: Did it say `$5.18`, not `518 cents` or the wrong currency?
- `pass@1` / `pass@2`: Did it succeed on the first try, or within two tries?
- `total_tool_calls`: Lower is better if correctness stays high.
- `max_turns_reached`: Bad. Means it got stuck/tool-looped.

Best current hard-holdout result:
- Reward: `0.835`
- Tool correctness: `1.000`
- Price format: `0.625`
- Avg tool calls: `1.812`

So: it is very good at placing/refusing orders correctly, but only medium at final receipt formatting.

</details>

**Other Qwen Sizes On Prime**

Prime currently lists these Qwen models:

<details>
<summary>👉 More info</summary>

| Model | Train cost / 1M tokens |
| --- | ---: |
| `Qwen/Qwen3.5-0.8B` | `$0.06` |
| `Qwen/Qwen3.5-2B` | `$0.15` |
| `Qwen/Qwen3.5-4B` | `$0.30` |
| `Qwen/Qwen3.5-9B` | `$0.60` |
| `Qwen/Qwen3.5-35B-A3B` | `$1.00` |
| `Qwen/Qwen3.5-122B-A10B` | `$2.00` |
| `Qwen/Qwen3.5-397B-A17B` | `$4.00` |
| `Qwen/Qwen3.6-35B-A3B` | `$1.00` |

Prices are per million training tokens.

</details>

**Should We Train Bigger Models?**

Not yet.

<details>
<summary>👉 More info</summary>

A bigger model may help, but our current failure is not mainly “the model is too dumb.” The main issue is that our reward/training signal for final receipts is too weak and sparse. We already proved the small model can learn the business rules very well.

Better next step:
- improve the training/eval examples around final receipt wording,
- make dollar-format grading stricter,
- then rerun a cheap 0.8B or maybe 2B/4B comparison.

Only train bigger if the small model plateaus after the grading signal is fixed.

</details>

**How Do We Decide If It Is Good Enough?**

<details>
<summary>👉 More info</summary>

I would call it good enough for a demo when:
- hard-holdout reward stays above `0.85`,
- tool correctness is `>= 0.98`,
- price format is `>= 0.95`,
- max-turn/tool-loop rate is near `0`,
- average tool calls stays around `<= 2`,
- manual samples look like real customer-facing receipts.

For production, I would also want:
- more holdout tasks,
- multi-item/follow-up order tests if we support those,
- regression tests against real app behavior,
- a human review of sampled conversations.

</details>

**What It Is Not Good At Yet**

<details>
<summary>👉 More info</summary>

Main weaknesses:
- It sometimes says prices as cents, like `518 cents`, instead of `$5.18`.
- It can still be more verbose than ideal.
- It sometimes does a menu lookup when it could place the order directly.
- We have not proven it handles multi-item orders or follow-up modifications.
- The hard holdout is still small, so we need more eval coverage before trusting it broadly.
</details>

Current verdict: we have a real improvement over vanilla Qwen, but the next work should tighten receipt formatting and concise confirmations before spending on larger models.

**Comment by kevinmichaelchen** [comment]
## ELI5: What We Did With Prime Intellect

We trained and tested a small coffee-shop ordering model like we were training a new cashier.

The goal was not just “make the model sound nicer.” The goal was:

| Goal | Meaning |
| --- | --- |
| Better than raw model | It should beat the plain Qwen model we started from. |
| Stay under budget | Use the $50 Prime credits carefully. |
| Real user value | The assistant should place/refuse orders correctly and give trustworthy receipts. |

<details>
<summary><strong>1. The Simple Story</strong></summary>

We started with a raw Qwen model. It was like a new cashier who sort of understands coffee orders but guesses too much.

We then built a custom grading environment for our coffee shop:

- Can it check the menu?
- Can it place a valid order with the right tool arguments?
- Can it refuse impossible orders?
- Can it stop after success instead of repeatedly using tools?
- Can it give a short customer-facing receipt?
- Can it write the total as dollars, like `$5.18`, not `518 cents`?

After several runs, the best model is still the Iteration 2 Qwen 0.8B adapter:

`rqvfrcdy4xt95oka37b0xjyu`

That adapter is much better than the raw model at operational coffee-ordering behavior.

</details>

<details>
<summary><strong>2. How We Grade The Model</strong></summary>

We use evals, which are like tests with answer keys.

| Metric | ELI5 Meaning |
| --- | --- |
| `tool_correctness` | Did it use the right tool with the right order details? |
| `final_response_quality` | Did the final answer mention the right receipt info? |
| `product_efficiency` | Did it avoid wasting tool calls or looping? |
| `price_format` | Did it say `$5.18` instead of `518 cents` or the wrong currency? |
| `receipt_style` | Was the receipt short, clean, and customer-facing? |
| `pass@1` / `pass@2` | Did it pass on the first try / within two tries? |

Current champion on the latest hard holdout:

| Metric | Score |
| --- | ---: |
| Reward avg | `0.830` |
| Pass@1 | `1.000` |
| Pass@2 | `1.000` |
| Tool correctness | `0.984` |
| Price format | `0.625` |
| Receipt style | `0.866` |
| Avg tool calls | `1.750` |

Translation: it is very good at the actual ordering behavior, but it still sometimes formats prices incorrectly.

</details>

<details>
<summary><strong>3. What Worked</strong></summary>

The biggest win was teaching the model coffee-shop tool discipline.

| Stage | Reward | Tool Correctness | Avg Tool Calls | Max-Turn Loop Rate |
| --- | ---: | ---: | ---: | ---: |
| Iteration 1 strict baseline | `0.703` | `0.953` | `3.000` | `0.562` |
| Iteration 2 champion | `0.907` | `1.000` | `1.812` | `0.000` |

ELI5: the model stopped acting like a cashier who keeps checking the menu over and over. It learned to place the order and stop.

</details>

<details>
<summary><strong>4. What Did Not Work</strong></summary>

We tried to fix receipt price formatting with more RL reward pressure. That did not beat the champion.

| Experiment | Result | Decision |
| --- | ---: | --- |
| 0.8B price hardening | `0.757` | Rejected |
| 0.8B warm-start price hardening | `0.747` | Rejected |
| 0.8B receipt-style warm-start | `0.743` | Rejected |
| Raw 2B | `0.610` | Not enough tool correctness |
| Raw 4B | `0.674` | Too verbose / truncated |
| 2B training probe | `0.668` | Stopped early to protect budget |

ELI5: bigger models were not automatically better cashiers. Some were better at writing prices, but worse at following the ordering process.

</details>

<details>
<summary><strong>5. Budget</strong></summary>

We started with `$50` in Prime credits.

| Item | Amount |
| --- | ---: |
| Current wallet | `$43.27` |
| Approx total spend | `$6.73` |
| Active training runs | `0` |
| Deployed adapters | `0` |

We stopped runs when they clearly would not beat the champion.

</details>

<details>
<summary><strong>6. Current Weakness</strong></summary>

The model’s remaining problem is narrow but important:

> It sometimes gives the right order but writes the price wrong.

Example of bad behavior:

```text
Latte ... 518 cents
```

Desired behavior:

```text
Latte order-simulated-0001 $5.18.
```

This matters because a coffee-shop assistant must give receipts customers can trust.

</details>

<details>
<summary><strong>7. Next Best Step</strong></summary>

The next best step is not another bigger RL run.

RL is like letting the model try many answers and grading them. That worked well for tool behavior, but receipt formatting is more like handwriting practice: we know exactly what the answer should look like.

So we prepared deterministic supervised fine-tuning style data:

| Artifact | Rows | Purpose |
| --- | ---: | --- |
| `receipt_final_response_sft.jsonl` | `22` | Teach final receipt wording directly. |
| `receipt_tool_trajectory_sft.jsonl` | `22` | Teach full tool-call + tool-result + receipt traces. |

These live under `.context/prime-intellect/lab/data/effect_coffee_sft/` for now.

The current Prime lab does not have the `uv run sft` entrypoint installed, so no SFT job was launched yet. The next step is either:

1. provision/install an SFT-capable Prime-RL runtime, or
2. convert the deterministic corpus into a small `receipt_drill` environment split if SFT remains unavailable.

</details>

**Comment by kevinmichaelchen** [comment]
Update after proceeding with the SFT path:

<details>
<summary>What changed</summary>

| Area | Result |
| --- | --- |
| Reproducible artifacts | Checked in the Prime experiment source under `experiments/prime-intellect/effect-coffee-ordering/`. |
| SFT data | Added deterministic receipt corpora: 22 prompt/completion rows and 22 tool-trajectory rows. |
| SFT runtime check | Prime-RL documents `uv run sft`, but the current Mac workspace cannot run it because the GitHub package resolves to CUDA Torch wheels with no macOS arm64 match. |
| Hosted SFT check | Current hosted training CLI exposes RL-style environment configs; hosted SFT is not launchable from here yet. |
| Fallback path | Added a `receipt_drill` environment split and a low-cost warmup config that starts from champion adapter `rqvfrcdy4xt95oka37b0xjyu`. |

</details>

<details>
<summary>ELI5 story</summary>

The model already knows how to use the coffee-shop tools pretty well. Its main mistake is how it talks after a good order: it sometimes says prices in the wrong format or keeps talking too much. So instead of paying for another broad training run, we made a small set of perfect examples that all say: after the order succeeds, give a tiny receipt like `Latte order-simulated-0001 $5.18.`

If Prime hosted SFT becomes available, we use those examples directly. If not, we now have a `receipt_drill` split that teaches the same lesson through the existing RL environment path.

</details>

<details>
<summary>Budget and gate</summary>

| Check | Current state |
| --- | --- |
| New paid training started | No |
| Running/Pending training runs | None found |
| Deployed adapters | None found |
| Wallet balance | `$43.2688` |
| Promotion rule | Beat champion hard-eval reward `0.830`, especially `price_format > 0.625`, without regressing tool correctness or verbosity. |

</details>

**Comment by kevinmichaelchen** [comment]
Follow-up: I published the receipt-drill environment snapshot as `kevinmichaelchen/effect-coffee-ordering@0.1.6` and pushed commit `69f47ca` so the checked-in `pyproject.toml` matches the registry version.

No training was launched. Current checks:

| Check | Result |
| --- | --- |
| Running/Pending training runs | none |
| Deployed adapters | none |
| Wallet balance | `$43.2688` |

The Prime environment action for the `0.1.6` publish is still running on Prime's side as of my latest poll; the version itself is visible as latest in `prime env status`.
